machida/lib/wallaroo/experimental/__init__.py

```diff
@@ -100,6 +125,325 @@ def write(self, message, event_time=0, key=None):
+        else:
+            self.producer_fifo.append(data)
+
+    def pending_sends(self):
+        """
+        Are there any pending sends
+        """
+        if len(self.producer_fifo) > 0:
+            return True
+        else:
+            return False
+
+    def notify(self, stream_id, stream_name=None, point_of_ref=None):
+        old = self._streams.get(stream_id, None)
+        if old:
+            if point_of_ref is None:
+                raise ConnectorError("Cannot update a stream without a valid "
+                                     "point_of_ref value")
+            new = Stream(stream_id,
+                         old.name if old is not None else stream_name,
+                         (point_of_ref if point_of_ref is not None else
+                          old.point_of_ref),
+                         old.is_open)
+        else:
+            if stream_name is None:
+                raise ConnectorError("Cannot notify a new stream without "
+                                     "a Stream name!")
+            new = Stream(stream_id,
+                         stream_name,
+                         0 if point_of_ref is None else point_of_ref,
+                         False)
+        # update locally
+        self._streams[stream_id] = new
+        # send to wallaroo worker
+        self.write(cwm.Notify(new.id,
+                              new.name,
+                              new.point_of_ref))
+
+    def end_of_stream(self, stream_id, event_time=None, key=None):
+        """
+        Send an EOS message for a stream_id, with an optional key and
+        event_time.
+        event_time must be either a datetime or a float of seconds since
+        epoch (it may be negtive for dates before 1970-1-1)
+        """
+        flags = cwm.Message.Eos | cwm.Message.Ephemeral
+        if event_time is not None:
+            flags |= cwm.Message.EventTime
+            if isinstance(event_time, datetime):
+                ts = dt_to_timestamp(event_time)
+            elif isinstance(event_time, (int, float)):
+                ts = event_time
+            else:
+                raise ProtocolError("Event_time must be a datetime or a float")
+        else:
+            ts = None
+        if key:
+            flags |= cw.Message.key
+            if isinstance(key, bytes):
+                en_key = key
+            else:
+                en_key = key.encode()
+        else:
+            en_key = None
+        msg = cwm.Message(
+            stream_id = stream_id,
+            flags = flags,
+            message_id = None,
+            event_time = ts,
+            key = en_key,
+            message = None)
+        self.write(msg)
+
+    ########################
+    # User defined methods #
+    ########################
+
+    def stream_updated(self, stream):
+        """
+        User handler can handle updates to their streams here.
+        Updates may include:
+            - stream has been closed (Stream.is_open = False)
```

Comment:
@jtfmumm suggests maybe breaking this up into `stream_closed`, `stream_opened`, and `point_of_ref_updated` methods

Reply:
point_of_ref_acked? 

Sidenote: the "point of reference" term was used to avoid confusing some generic idea of reference in some contexts like the Erlang spec but in many cases there is no ambiguity. I'd be fine using the word reference on it's own but I'm not sure if others would find that too ambiguous in places like this.

Reply:
I really like `point_of_ref` / "point of reference". I think it's both clear and disambiguates cases where the direct message id isn't the actual one.

I don't know if `point_of_ref_acked` is better than `updated`? What's the action on `acked`? Could we skip calling into this function if the point_of_ref was acked but to the same value we already had in our state?